apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -0,0 +1,315 @@
+-module(dcos_dns_handler).
+
+-include("dcos_dns.hrl").
+-include_lib("dns/include/dns_records.hrl").
+
+-define(TIMEOUT, 5000).
+
+%% API
+-export([
+    start/3,
+    resolve/3
+]).
+
+%% Private functions
+-export([start_link/3, init/4]).
+
+-export_type([reply_fun/0, protocol/0]).
+
+-type reply_fun() :: fun ((binary()) -> ok | {error, term()})
+                   | {fun ((...) -> ok | {error, term()}), [term()]}.
+-type protocol() :: udp | tcp.
+
+-spec(start(protocol(), binary(), reply_fun()) ->
+    {ok, pid()} | {error, overload}).
+start(Protocol, Data, Fun) ->
+    Args = [Protocol, Data, Fun],
+    case sidejob_supervisor:start_child(dcos_dns_handler_sj,
+                                        ?MODULE, start_link, Args) of
+        {error, overload} ->
```

Comment:
is this the only error possible? Asking because we are matching it with the exact atom.